File: internal/message/subscribe.go
package message

import (
	"io"
	"log/slog"
	"time"

	"github.com/quic-go/quic-go/quicvarint"
)

type SubscribeID uint64

type TrackAlias uint64

type TrackPriority byte

type GroupOrder byte

type SubscribeMessage struct {
	SubscribeID SubscribeID

	TrackPath string

	TrackPriority TrackPriority

	/*
	 * The order of the group
	 * This defines how the media is played
	 */
	GroupOrder GroupOrder

	GroupExpires time.Duration

	/***/
	MinGroupSequence GroupSequence

	/***/
	MaxGroupSequence GroupSequence

	/*
	 * Subscribe Parameters
	 * Parameters should contain Authorization Information
	 */
	Parameters Parameters
}

func (s SubscribeMessage) Encode(w io.Writer) error {
	slog.Debug("encoding a SUBSCRIBE message")

	/*
	 * Serialize the message in the following formatt
	 *
	 * SUBSCRIBE Message payload {
	 *   Track Path (string),
	 *   Subscriber Priority (8),
	 *   Group Order (8),
	 *   Group Expires (varint),
	 *   Min Group Sequence (varint),
	 *   Max Group Sequence (varint),
	 *   Subscribe Parameters (Parameters),
	 * }
	 */

	/*
	 * Serialize the payload
	 */
	p := make([]byte, 0, 1<<6)

	// Append the Subscriber ID
	p = quicvarint.Append(p, uint64(s.SubscribeID))

	// Append the Track Name
	p = quicvarint.Append(p, uint64(len(s.TrackPath)))
	p = append(p, []byte(s.TrackPath)...)

	// Append the Subscriber Priority
	p = append(p, []byte{byte(s.TrackPriority)}...)

	// Append the Group Order
	p = append(p, []byte{byte(s.GroupOrder)}...)

	// Append the Group Expires
	p = append(p, []byte{byte(s.GroupOrder)}...)

	// Append the Min Group Sequence
	p = quicvarint.Append(p, uint64(s.MinGroupSequence))

	// Append the Max Group Sequence
	p = quicvarint.Append(p, uint64(s.MinGroupSequence))

	// Append the Subscribe Update Priority
	p = appendParameters(p, s.Parameters)

	// Get a serialized message
	b := make([]byte, 0, len(p)+8)

	// Append the length of the payload
	b = quicvarint.Append(b, uint64(len(p)))

	// Append the payload
	b = append(b, p...)

	// Write
	_, err := w.Write(b)
	if err != nil {
		slog.Error("failed to write a SUBSCRIBE message", slog.String("error", err.Error()))
		return err
	}

	slog.Debug("encoded a SUBSCRIBE message")

	return err
}

func (s *SubscribeMessage) Decode(r io.Reader) error {
	slog.Debug("decoding a SUBSCRIBE message")

	// Get a messaga reader
	mr, err := newReader(r)
	if err != nil {
		return err
	}

	// Get Subscribe ID
	num, err := quicvarint.Read(mr)
	if err != nil {
		return err
	}
	s.SubscribeID = SubscribeID(num)

	// Get Track Name
	num, err = quicvarint.Read(mr)
	if err != nil {
		return err
	}
	buf := make([]byte, num)
	_, err = r.Read(buf)
	if err != nil {
		return err
	}
	s.TrackPath = string(buf)

	// Get Subscriber Priority
	bnum, err := mr.ReadByte()
	if err != nil {
		return err
	}
	s.TrackPriority = TrackPriority(bnum)

	// Get Group Order
	bnum, err = mr.ReadByte()
	if err != nil {
		return err
	}
	s.GroupOrder = GroupOrder(bnum)

	// Get Group Expires
	num, err = quicvarint.Read(mr)
	if err != nil {
		return err
	}
	s.GroupExpires = time.Duration(num)

	// Get Min Group Sequence
	num, err = quicvarint.Read(mr)
	if err != nil {
		return err
	}
	s.MinGroupSequence = GroupSequence(num)

	// Get Max Group Sequence
	num, err = quicvarint.Read(mr)
	if err != nil {
		return err
	}
	s.MaxGroupSequence = GroupSequence(num)

	// Get Subscribe Update Parameters
	s.Parameters, err = readParameters(mr)
	if err != nil {
		return err
	}

	slog.Debug("decoded a SUBSCRIBE message")

	return nil
}

File: interop/web/src/message/subscribe.ts
import { Writer, Reader } from "../io";
import { varintLen, stringLen } from "../io/len";

export class SubscribeMessage {
    subscribeId: bigint;
    broadcastPath: string;
    trackName: string;
    trackPriority: bigint;
    minGroupSequence: bigint;
    maxGroupSequence: bigint;

    constructor(subscribeId: bigint, broadcastPath: string, trackName: string,
         trackPriority: bigint, minGroupSequence: bigint, maxGroupSequence: bigint) {
        this.subscribeId = subscribeId;
        this.broadcastPath = broadcastPath;
        this.trackName = trackName;
        this.trackPriority = trackPriority;
        this.minGroupSequence = minGroupSequence;
        this.maxGroupSequence = maxGroupSequence;
    }

    length(): number {
        return (
            varintLen(this.subscribeId) +
            stringLen(this.broadcastPath) +
            stringLen(this.trackName) +
            varintLen(this.trackPriority) +
            varintLen(this.minGroupSequence) +
            varintLen(this.maxGroupSequence)
        );
    }


    static async encode(writer: Writer, subscribeId: bigint, broadcastPath: string, trackName: string,
         trackPriority: bigint, minGroupSequence: bigint, maxGroupSequence: bigint): Promise<[SubscribeMessage?, Error?]> {
        const msg = new SubscribeMessage(subscribeId, broadcastPath, trackName, trackPriority, minGroupSequence, maxGroupSequence);
        writer.writeVarint(BigInt(msg.length()));
        writer.writeVarint(subscribeId);
        writer.writeString(broadcastPath);
        writer.writeString(trackName);
        writer.writeVarint(trackPriority);
        writer.writeVarint(minGroupSequence);
        writer.writeVarint(maxGroupSequence);
        const err = await writer.flush();
        if (err) {
            return [undefined, err];
        }
        return [msg, undefined];
    }

    static async decode(reader: Reader): Promise<[SubscribeMessage?, Error?]> {
        const [len, err] = await reader.readVarint();
        if (err) {
            return [undefined, new Error("Failed to read length for SubscribeMessage: " + err.message)];
        }

        const [subscribeId, err2] = await reader.readVarint();
        if (err2) {
            return [undefined, new Error("Failed to read subscribeId for SubscribeMessage")];
        }

        const [broadcastPath, err3] = await reader.readString();
        if (err3) {
            return [undefined, new Error("Failed to read broadcastPath for SubscribeMessage")];
        }

        const [trackName, err4] = await reader.readString();
        if (err4) {
            return [undefined, new Error("Failed to read trackName for SubscribeMessage")];
        }

        const [trackPriority, err5] = await reader.readVarint();
        if (err5) {
            return [undefined, new Error("Failed to read trackPriority for SubscribeMessage")];
        }

        const [minGroupSequence, err6] = await reader.readVarint();
        if (err6) {
            return [undefined, new Error("Failed to read minGroupSequence for SubscribeMessage")];
        }

        const [maxGroupSequence, err7] = await reader.readVarint();
        if (err7) {
            return [undefined, new Error("Failed to read maxGroupSequence for SubscribeMessage")];
        }

        return [new SubscribeMessage(subscribeId, broadcastPath, trackName, trackPriority, minGroupSequence, maxGroupSequence), undefined];
    }
}